vault-plugins/flant_iam_auth/usecase/authn/accesstoken/validator.go

```diff
@@ -0,0 +1,319 @@
+package accesstoken
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"mime"
+	"net/http"
+	"strings"
+	"time"
+
+	hcjwt "github.com/hashicorp/cap/jwt"
+	"github.com/hashicorp/go-cleanhttp"
+	"golang.org/x/oauth2"
+	"gopkg.in/square/go-jose.v2/jwt"
+
+	"github.com/flant/negentropy/vault-plugins/flant_iam_auth/model"
+	jwt2 "github.com/flant/negentropy/vault-plugins/flant_iam_auth/usecase/authn/jwt"
+)
+
+// DefaultLeewaySeconds defines the amount of leeway that's used by default
+// for validating the "nbf" (Not Before) and "exp" (Expiration Time) claims.
+const DefaultLeewaySeconds = 150
+
+type accessTokenValidator struct {
+	userinfoURL string
+	issuer      string
+	client      *http.Client
+}
+
+func NewAccessTokenValidator(ctx context.Context, cfg *model.AuthSource) (jwt2.JwtValidator, error) {
+	if cfg == nil || cfg.OIDCDiscoveryURL == "" {
+		return nil, fmt.Errorf("%v", cfg)
+	}
+	caCtx, err := createCAContext(ctx, cfg.OIDCDiscoveryCAPEM)
+	if err != nil {
+		return nil, err
+	}
+	client := http.DefaultClient
+	if c, ok := caCtx.Value(oauth2.HTTPClient).(*http.Client); ok {
+		client = c
+	}
+
+	wellKnown := strings.TrimSuffix(cfg.OIDCDiscoveryURL, "/") + "/.well-known/openid-configuration"
+	req, err := http.NewRequest(http.MethodGet, wellKnown, nil)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := client.Do(req.WithContext(caCtx))
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	// Read the response body and status code
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read response body: %v", err)
+	}
+	if resp.StatusCode != http.StatusOK {
```

Comment:
maybe check status first and after read body?

Reply:
in case not OK - we returns the body to better information